File: Sean的leetcode/Leetcode/Palindrome Linked List.cpp
//https://leetcode.com/discuss/44733/c-o-n-time-o-1-space-solution
//reverse the 2nd half list then compare with the 1st half list.
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *reverse(ListNode *head)
    {
        ListNode *pre = 0;
        ListNode *curr = head;

        while (curr)
        {
            ListNode *next = curr->next;
            curr->next = pre;
            pre = curr;
            curr = next;
        }
        return pre;
    }

    bool isPalindrome(ListNode* head) {
        ListNode *curr = head;
        ListNode *mid;
        int num = 0;

        if (!head) return true;

        while (curr)
        {
            num++;
            curr = curr->next;
        }

        curr = head;
        for (int i = 0; i < num >> 1; i++)
            curr = curr->next;

        mid = reverse(curr);
        curr = head;
        while (mid)
        {
            if (mid->val != curr->val)
                return false;
            mid = mid->next;
            curr = curr->next;
        }
        return true;
    }
};

File: MyLeetCode/python/Palindrome Linked List.py
###
# Time Complexity: O(n)
# Space Complexity: O(1)
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        if not head:
            return True

        l1_end = self.find_mid(head)
        l2 = self.reverse(l1_end)

        p1 = head
        p2 = l2
        while p1 and p2:
            if p1.val != p2.val:
                return False
            p1 = p1.next
            p2 = p2.next
        return True

    def find_mid(self, head):
        s = f = head
        while f and f.next:
            f = f.next.next
            s = s.next
        return s

    def reverse(self, head):
        pre = None
        p = head
        while p:
            nxt = p.next
            p.next = pre
            pre = p
            p = nxt
        return pre
